Unit Tests

Guiorgy opened the pull request:
### TODO:
- [ ] ReadOnlySpan Tests
  - [ ] Linq Tests
    - todo
  - [ ] Split Tests
    - [x] Fuzzing Tests
      - [x] Array Source
        - [x] Single Delimiter
        - [x] Single Delimiter With Count and CountExceedingBehaviour
        - [x] Any Delimiter
        - [x] Any Delimiter With Count and CountExceedingBehaviour
        - [x] Delimiter Sequence
        - [x] Delimiter Sequence With Count and CountExceedingBehaviour
      - [x] String Source
        - [x] Single Delimiter With StringSplitOptions
        - [x] Single Delimiter With Count, StringSplitOptions and CountExceedingBehaviour
        - [x] Any Delimiter With StringSplitOptions
        - [x] Any Delimiter With Count, StringSplitOptions and CountExceedingBehaviour
        - [x] Delimiter Sequence With StringSplitOptions
        - [x] Delimiter Sequence With Count, StringSplitOptions and CountExceedingBehaviour
    - [ ] Facts
      - [x] Enumeration Returns the Correct Span Type
      - [x] Empty Source Results in Empty Span
      - [x] Consecutive Delimiters Result in an Empty Span
      - [x] Delimiter at the Start Results in an Empty Span
      - [x] Delimiter at the Start With RemoveEmptyEntries Option Doesn't Result in an Empty Span
      - [x] Delimiter at the End Results in an Empty Span
      - [x] Delimiter at the End With RemoveEmptyEntries Option Doesn't Result in an Empty Span
      - [x] Default CountExceedingBehaviour Option is AppendLastElements
      - [x] Delimiter at the End With Count Equal to Delimiter Count Results in a Span at the End With a Delimiter
      - [x] By Default Count Equal to Delimiter Count Results in Span at the End With Everything After and Including Last Delimiter
      - [x] Count Equal to Delimiter Count with CutLastElements Option Results in Everything After and Including Last Delimiter Being Cut
      - [x] Consecutive Delimiters With RemoveEmptyEntries Option Doesn't Result in an Empty Span
      - [x] Consecutive Delimiters at the End With Count Equal to Delimiter Count With RemoveEmptyEntries Option Doesn't Result in a Span With Delimiter
      - [x] Trim Entries Option Should Trim Last Span
      - [x] White-Space Span With TrimEntries and RemoveEmptyEntries Options Should Return Nothing
      - [x] Consecutive Delimiters at the End With RemoveEmptyEntries Option Doesn't Result in Empty Spans
      - [x] Count Equal to 0 Returns Nothing
      - [x] Negative Count Throws an ArgumentOutOfRangeException
      - [x] Count Equal 1 With RemoveEmptyEntries Option Returns Nothing if Source is Empty
      - [x] Count Equal 1 With RemoveEmptyEntries Option Doesn't Recursively Remove Empty Spans
      - [x] Count Greater Than 1 With RemoveEmptyEntries Option Recursively Removes Empty Spans
      - [x] Count Equal 1 With RemoveEmptyEntries and TrimEntries Options Returns Nothing if Source is White Space
      - [x] Count Equal 1 With RemoveEmptyEntries and TrimEntries Options Doesn't Recursively Remove White Space Spans
      - [x] Count Greater Than 1 With RemoveEmptyEntries and TrimEntries Options Recursively Removes White Space Spans
      - [x] Undefned CountExceedingBehaviour Option Throws an ArgumentException
      - [x] Undefned StringSplitOptions Option Throws an ArgumentException
      - [x] Delimiter Span With Length Zero Should Return Source Span Unchanged
      - [x] Delimiter Span With Length Zero Should Assume Count is Equal One
  - [ ] String Tests
    - todo
- [ ] Span Tests
  - [ ] Linq Tests
    - todo
  - [ ] Split Tests
    - [ ] Same as ReadOnlySpan Tests
  - [ ] String Tests
    - todo

dragon7307 commented:
We probably would first need to agree on a testing framework to use and this is when some creative differences will come into play. So, I personally use **XUnit** for my tests, since I find it the most intuitive framework.  You seem to be using MSTest if I am not mistaken - a testing framework I've never used.   

As a quick note, I would also rename the folder `Tests` to `tests for consistency in naming with `src`.

Guiorgy commented:
I don't really have much of a preference, MSTest is just what I've defaulted to use. I haven't personally used XUnit, but can't imagine it being much different, so I'll just migrate the project.

I'll also change the name no problem.

More importantly, while writing tests for `ReadOnlySpan<char>`, I assumed you were trying to make the `Split(char, StringSplitOprions, int)` methods compatible to `string.Split(char, int, StringSplitOprions)`. If this is the case, then there's a problem.  The inmplementation in string, as the [docs](https://learn.microsoft.com/en-us/dotnet/api/system.string.split?view=net-8.0#system-string-split(system-char-system-int32-system-stringsplitoptions)) state, _**if the string has already been split count - 1 times, but the end of the string has not been reached, then the last string in the returned array will contain this instance's remaining trailing substring, untouched.**_. Your implementation on the other hand cuts at the `count`-th delimiter. For example:

```cs
string @string = new(['2', 'x', 'r', 'b', '7', '6', 'p', 'w', 'd', 'r', '6', '4', 'g', 't', 'i', '6', '5', 'c', 'm', 'r', 't', '4', '7', 'v', '3', 'y', '6', '5', 'n', 'b', 'n', 't', 'p', '4', 'r', 'w', '6', '1', 'x', 'a', 't', 'i', 'i', 'o', 'o', 'q', '4', '0', 'q', 'i', '6', 'd', 'x', 'i', '9', '6', 'd', 'w', 'e', '2', 'x', 's', 'a', '9', 't', 'g', 'p', 'o', 'k', 'o', 'm', 'e', 'p', '5', 'k', '7', 's', '3', '0', 'c', 's', 'z', 'b', '2', 'j', '5', 'b', 's', 'u', 'h', 'c', '8', 'i', 'w', 'y', 'c', '8', 'u', 'f', 'f']);
char delimiter = 'o';
ReadOnlySpan<char> charSpan = @string;

var expected = string.Join(", ", @string.Split(delimiter, 3, StringSplitOptions.None));
var actual = string.Join(", ", ToSystemEnumerable(charSpan.Split(delimiter, StringSplitOptions.None, 3)).Select(a => string.Join("", a)));

/*
Results:
string: 2xrb76pwdr64gti65cmrt47v3y65nbntp4rw61xatiiooq40qi6dxi96dwe2xsa9tgpokomep5k7s30cszb2j5bsuhc8iwyc8uff
expexted: 2xrb76pwdr64gti65cmrt47v3y65nbntp4rw61xatii, , q40qi6dxi96dwe2xsa9tgpokomep5k7s30cszb2j5bsuhc8iwyc8uff
actual: 2xrb76pwdr64gti65cmrt47v3y65nbntp4rw61xatii, , q40qi6dxi96dwe2xsa9tgp
*/
```

dragon7307 commented:
This is interesting, I was not aware `string.Split` with `count` is working, that way. This is a design consideration for which I'll need some time to settle on a conclusion.   Not quite sure yet, what the solution should be, maybe one could pass an optional boolean parameter to determine, whether or not the end shouldz be appended or cut..
I'll try both approaches and will hopefully be back with a conclusion.

Guiorgy commented:
For now I'll write them assuming compatibility with `string.Split`, we can change this later

Guiorgy commented:
The reason why empty sequences are dropped when the delimiter is at the end of the source sequence, is because of the assumption that the remaining span is empty only when it's been exhausted:

```cs
public bool MoveNext()
{
    ReadOnlySpan<T> span = Span;
    if(span.IsEmpty) // wrong
    {
        return false;
    }
    int index = span.IndexOf(Delimiter);

    if(index == -1 || index >= span.Length)
    {
        Span = ReadOnlySpan<T>.Empty; // 1 way it can become empty
        Current = span;
        return true;
    }
    Current = span[..index];
    Span = span[(index + 1)..]; // second way it can become empty
    return true;
}
```

We probably have to store an extra flag to indicate the end:

```cs
private bool done;
public bool MoveNext()
{
    ReadOnlySpan<T> span = Span;
    if(done)
    {
        return false;
    }
    int index = span.IndexOf(Delimiter);

    if(index == -1 || index >= span.Length)
    {
        done = true; // only here it can become true
        Current = span;
        return true;
    }
    Current = span[..index];
    Span = span[(index + 1)..]; // done will always be false here
    return true;
}
```

The name can be agreed on.

Guiorgy commented:
It's a bit unfortunate that those tests for `ReadOnlySpan` and `Span` are basically exact copies of each other. In the future if there's time, might consider making a source generator that takes tests for `ReadOnlySpan` and just replaces `ReadOnlySpan` with `Span`, if you are interested in that :P

dragon7307 commented:
So, 
In fact I was trying to make the `Span` and `String` Methods at least similiar - I have swapped the order of the `count` and the `stringSplitOptions` parameters on the **dev**-branch.   
I've fixed the problem with the end of the `Span` being cut off after `count` has been reached, by implementing an _enum_ called `CountExceedingBehaviour`, which is passed as an optional argument to the `count` methods, and is defaulted to `CountExceedingBehaviour.AppendLastElements`, which essentially mirrors the `string` version. Optionally one can get the old behaviour back with the cut, by passing `CountExceedingBehaviour.CutLastElements`.  The fix unfortunately requires concatenating two `Spans`, what I am currently doing with an intermediate `array`. This however is not an ideal solution since the enumerators are supposed to be allocation free - so I'll be looking for an optimized version in the future.   

A Source Generator really sounds like something to investigate in. That will be the future though.   
For now I'll resolve the merge conflicts in your PR... 
 
Also I unfortunately won't be available tomorrow, e.g. in the next 24 hours, but after that I'll be going through the test suite  and all the other concerns.

dragon7307 commented:
Another curious issue I found, is that when splitting a span consisting of delimiters only, and specifiyng the `length` of the span as `count`, it returns `count` blank spans.  However the `string` implementation returns `count - 1` blank lines and one time the `delimiter `itself.

Guiorgy commented:
The tests aren't done, I've only implemented ones for the `Split` and `SplitAny` methods. I'll also be updating them with the new changes.. I'll tell you when they are ready to be merged.

I noticed that you swapped `StringSplitOptions` and `count`, so I've already updated those tests.

From what I can see, as I mentioned in my comment under the commit, the default behaviour is `CountExceedingBehaviour countExceedingBehaviour = CountExceedingBehaviour.CutLastElements`, so the old behaviour, not the new one.

dragon7307 commented:
Right, It is indeed. The behaviour shouldn't be that way though. I'll make the new one the default.

Guiorgy commented:
About the string split issue you mentioned, the string implementation is correct. Here's an example:

```cs
var src = "aaa"; // when split by 'a' we get ['', 'a', '', 'a', '', 'a', ''] 4 empty around 3 'a'-s
src.Split('a'); // ['', '', '', ''] because 
src.Split('a', 4); // ['', '', '', '']
src.Split('a', 3); // ['', '', 'a'] where 'a' is '' + 'a' + '' (3rd empty and after)
src.Split('a', 2); // ['', 'aa''] where 'aa' is '' + 'a' + '' + 'a' + '' (second empty and after)
```

or another example to better visualize:

```cs
var src = "bababab";
src.Split('a'); // ['b', 'b', 'b', 'b']
src.Split('a', 4); // ['b', 'b', 'b', 'b']
src.Split('a', 3); // ['b', 'b', 'bab']
src.Split('a', 2); // ['b', 'babab'']
```

Guiorgy commented:
Thinking about it some more, this assumption is also wrong when `options` is `TrimEntries`:

```cs
ReadOnlySpan<char> span = Span;
int index = span.IndexOf(Delimiter);

if(index == -1)
{
    enumerationDone = true;
    Current = span;
    return true;
}
```

Since this shortcircuits, it never checks if trimming is necessary, for example:

```cs
" a b a ".AsSpan().Split('b', StringSplitOptions.TrimEntries);
```

results in `[['a'], [' ', 'a', ' ']]` (last span was **not** trimmed).

I'll rewrite the methods and include comments if you are ok with that.

Guiorgy commented:
Turns out you can have a Draft PR (didn't know :D) that can't be accidentally merged until ready.

dragon7307 commented:
:D

Guiorgy commented:
I run into another weird differebce:

```cs
var str = "abcdababcd";

Console.WriteLine(str.Split("ab", 1, StringSplitOptions.RemoveEmptyEntries)); // abcdababcd
Console.WriteLine(str.AsSpan().Split("ab", 1, StringSplitOptions.RemoveEmptyEntries)); // cdababcd
Console.WriteLine(str.Split("ab", 2, StringSplitOptions.RemoveEmptyEntries)); // cd, cd
Console.WriteLine(str.AsSpan().Split("ab", 2, StringSplitOptions.RemoveEmptyEntries)); // cd, cd
```

What's weir is that when `count` is more than 1, and `RemoveEmptyEntries` is set, on the last split it will remove all preceding delimiters:
1. abcdababcd
2. '' - cdababcd
  2.1. remove empty: cdababcd
3. cd - abcd
  3.1. last split, remove preciding delimiters: cd, '' - cd = cd, cd

But when initial `count` is 1, it ignores that and just returns the string as is, even though I'd expect it to do the same:
1. abcdababcd
  1.1. last split, remove preciding delimiters: '' - cdababcd = cdababcd

dragon7307 commented:
I agree, this seems oddly inconsistent.
That however raises the question once again, whether it is even a good idea to try to mimick `string.Split` or whether it would be better to allow some creative differences for the better.

Guiorgy commented:
Whether we should try to mimick `string.Split` depends on the goal, is it to be a drop in replacement to exising (`string.Split`) methods that are (hopefuly) faster (needs testing, my next goal are benchmarks), or is it something else? If it's the former, then we should try to make the behaviour identical, at least by default. It's fine to add optional features. At least, that's my opinion.

dragon7307 commented:
Wait with the benchmarking for now. I have already started writing benchmarks on a local branch, which I will push, and the `Span.Split` methods are faster by quite a margin and more importantly allocation-free - well mostly, since the behaviour for the appendlastelements option, currently requires some allocation. I am working on fixing that.

Guiorgy commented:
I gave the benchmarks you wrote a quick look, and personally, I would've generated the source array/string and delimiter(s) with a constant seed to test relative speeds in different conditions, for example:

- 10 char string with
  - no delimiter(s) occurrence
  - 1 delimiter(s) occurrence
  - 1/3 characters are delimiters
- 100 char string with
  - no delimiter(s) occurrence
  - 10 delimiter(s) occurrence
  - 1/3 characters are delimiters
- 1000 char string with
  - no delimiter(s) occurrence
  - 100 delimiter(s) occurrence
  - 1/3 characters are delimiters

This way we can see how the performance scales with the size of the source sequence, and the number of delimiter occurrences.

dragon7307 commented:
Not a bad call, but the thing for me is, that, as you can see from running the benchmarks, the performace differs based on factors like location of the delimiter. With pseudo-randomly generated sequences there would be no control about such things.
I agree with you, that the current data for benchmarking is not ideal, but this is something to work on, when the implementation is done.  
However I would like to keep the iterator based approach for the parameters to be able to exercise a certain degree of control over the data.  Scaling is definitely a factor to be considered.  But this can be implemented into the iterator.  
 
Unless you have a convincing reason to change the approacb I would tend to keep the current. 
Open for suggestions.

Guiorgy commented:
As far as I understand, the fluctuations in delimiter location are and such are exactly why you'd want a big random data set to get something closer to the average. We wouldn't generate just 1 test data for each case, but instead a large amount, for example `10000` strings and delimiter combos (or maybe even more), and run the relevant method in a loop, this way any such fluctuations should be averaged out.

For example:

```cs
public class ReadOnlySpan_Split_Benchmark
{
    [Benchmark]
    [ArgumentsSource(nameof(GetDataSets))]
    public int StringAsSpanSplit((string value, char delimiter)[] dataset)
    {
        int sum = 0;
        foreach((string value, char delimiter) in dataset)
        {
            foreach(ReadOnlySpan<char> split in value.AsSpan().Split(delimiter))
            {
                sum += split.Length;
            }
        }
        return sum;
    }

    [Benchmark(Baseline = true)]
    [ArgumentsSource(nameof(GetDataSets))]
    public int StringSplit((string value, char delimiter)[] dataset)
    {
        int sum = 0;
        foreach((string value, char delimiter) in dataset)
        {
            foreach(string split in value.Split(delimiter))
            {
                sum += split.Length;
            }
        }
        return sum;
    }

    static readonly Random random = new Random(123456789); // fixed seed to have comparable benchmarks

    static string GenerateRandomString(int length)
    {
        const string alphabet = "abcdefghijklmnopqrstuvwxyz0123456789 ";

        StringBuilder builder = new(length);
        for(int i = 0; i < length; i++)
        {
            int alphabetIndex = random.Next(alphabet.Length);
            builder.Append(alphabet[alphabetIndex]);
        }

        return builder.ToString();
    }

    public sealed class DataSet((string value, char delimiter)[] data) : IEnumerable
    {
        public readonly (string value, char delimiter)[] data = data;

        public static implicit operator DataSet((string value, char delimiter)[] data) => new(data);
        public static implicit operator (string value, char delimiter)[](DataSet dataset) => dataset.data;

        public IEnumerator GetEnumerator()
        {
            return data.GetEnumerator();
        }

        public override string ToString()
        {
            return $"Length: {data[0].value.Length}";
        }
    }

    public static IEnumerable<DataSet> GetDataSets()
    {
        const int datasetSize = 10000;

        int stringLength = 10;
        var dataset = new (string value, char delimiter)[datasetSize];
        for(int i = 0; i < datasetSize; i++)
        {
            string value = GenerateRandomString(stringLength);
            char delimiter = value[random.Next(value.Length)];
            dataset[i] = (value, delimiter);
        }
        yield return dataset;

        stringLength = 100;
        dataset = new (string value, char delimiter)[datasetSize];
        for(int i = 0; i < datasetSize; i++)
        {
            string value = GenerateRandomString(stringLength);
            char delimiter = value[random.Next(value.Length)];
            dataset[i] = (value, delimiter);
        }
        yield return dataset;

        stringLength = 1000;
        dataset = new (string value, char delimiter)[datasetSize];
        for(int i = 0; i < datasetSize; i++)
        {
            string value = GenerateRandomString(stringLength);
            char delimiter = value[random.Next(value.Length)];
            dataset[i] = (value, delimiter);
        }
        yield return dataset;
    }
}
```

Note: I didn't control the number of delimiter ocurences in the string this time, since it's an example.

Results:

| Method            | dataset      | Mean       | Error    | StdDev   | Median     | Ratio | Allocated  | Alloc Ratio |
|------------------ |------------- |-----------:|---------:|---------:|-----------:|------:|-----------:|------------:|
| StringAsSpanSplit | Length: 10   |   167.8 us |  2.43 us |  2.03 us |   167.7 us |  0.39 |          - |        0.00 |
| StringSplit       | Length: 10   |   418.5 us |  7.36 us | 16.16 us |   410.4 us |  1.00 |  1074560 B |        1.00 |
|                   |              |            |          |          |            |       |            |             |
| StringAsSpanSplit | Length: 100  |   468.8 us |  6.39 us |  5.97 us |   468.8 us |  0.37 |          - |        0.00 |
| StringSplit       | Length: 100  | 1,280.4 us | 12.60 us | 11.17 us | 1,278.0 us |  1.00 |  3660440 B |        1.00 |
|                   |              |            |          |          |            |       |            |             |
| StringAsSpanSplit | Length: 1000 | 3,421.7 us | 13.41 us | 12.54 us | 3,421.4 us |  0.42 |          - |        0.00 |
| StringSplit       | Length: 1000 | 8,195.3 us | 88.81 us | 78.72 us | 8,205.6 us |  1.00 | 29022169 B |        1.00 |

dragon7307 commented:
Probably should have mentioned that I want a small benchmark suite to run, as well as a portion of the unit tests, whenever a pull request is opened or there is a push to the main branch, Essentially I want a failing test on PRs or Commits that decrease Performance. I am not sure how reliable github actions performance is in this regard or how high the standard deviation will be..

However what you are proposing seems like a good idea for extensive testing, maybe before releases. Maybe it could be implemented as separate benchmarks for a different purpose.

Guiorgy commented:
I see. My goal for benchmarks was to have a comprehensive comparison against existing solutions as well as writing things manually (in case of split for example, calling `IndexOf` and `Slice` manually`) to show how much one can gain agains exising solutions, and how much overhead there is agains doing it by hand. It was something meant to be run once in every few updates to display in the readme, or a separate `benchmarks.md` file that would be referenced in the readme. Basically, marketing to those considering using the library.

Since your goal is different, might be worth doing as you stated and having 2 separate projects for that?

dragon7307 commented:
Sounds like a good solution to me, more Benchmarks means more precise results.

Guiorgy commented:
Might be worth having a look at how [.NET](https://github.com/dotnet/performance) does performance testing for your case, though I haven't looked myself.

Guiorgy commented:
I've been thinking about the continuous performance regression tests you said you wanted, and there are a few things to consider.

There's probably no guarantee that benchmarks run previously and now were on the same hardware configuration, let alone same machine. For example, it might've been an Intel 12th gen server yesterday, it could be an AMD Epyc server today. Results may vary substantially. Relying on that won't work in my opinion.

If possible, we'd need benchmarks to be run on the last known commit AND the commit that we wish to test for performance regression every time, so hopefully they are run in same conditions, so the results can be compared. I don't know if this is possible, since I've never touched GitHub actions :P

Then even if the hardware and conditions are the same, we need to make sure the tests are very consistent.

Back to unit tests, I've listed all currently implemented tests in the PR description, if there are any test cases you wish to see you can list them here, I'll add them to the todo list.

dragon7307 commented:
So as we decided to create two projects for benchmarking and probably two for unit testing, I would recommend we have a folde structure somewhat like this:   

---src
---tests
    ---Correctness
        ---Minimal
        ---Extended
    ---Performance
        ---Minimal
        ---Extended
         

The Minimal projects are supposedu to be used in automation via Github Actions  and will be run on each pull request opened, the Extended projects are run on release.  

 
Also I perked into your tests and even though I have only thrown a quick glance at it, shouldn't the `Random` in TestHelper be seeded??

Guiorgy commented:
For benchmarks you want seeded, but for unit tests the opposite. There's always a chance that the unit tests will catch a case we didn't account for by sheer luck, most of the ones I found were found like that :D

That also means that if a random test fails, it needs to be seen if this is a new uncounted case, or a regression, so we should decide if we want them to be used to fail commits.

In fact, since every time the (random) case is run the data is different, is exactly why I also added a method that generates a message with all the test data when a test fails, so that the case can be reproduced.

dragon7307 commented:
Well actually I am not sure if you want unit tests that are meant to be run for production  to be random since this of course means that a build may fail even though it was the same.  Randomized Unit Tests sounds to me more like something you want to run once on a huge scale, but unit tests for production are to make sure that no new bugs were introduced. Randomized unit tests are dangerous because yes they can find new edge cases, but they can also miss bugs that were introduced, because maybe they didn't cover a particular aspect in one run they covered in all the other runs..  So no, I do not agree, that unit tests should be randomized, at least not if one intends to run them before releasing a new version. New Bugs could go unnoticed. Tests that find new cases are useful, for testing the entire framework, but not to test prod code. Unit tests need to protect against regression primarily. I would not say, we should not have randomized tests, I am just saying that we do need non-random unit tests for production. 

Microsoft even list Repeatability as a factor of a good unit test  (https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices#characteristics-of-a-good-unit-test).

Guiorgy commented:
There seems to have been a misunderstanding between us. I also agree that unit tests meant for integration testing and regression testing should be deterministic. The tests I weote are divided into two groups, `Random` and `Facts`. Facts are predefined deterministic (with 0 randomness) for regression testing. The random tests are only meant as a bonus to discover new edge cases. Every time a new one apears, it ahould be added to the Facts (that's where most of the ones I wrote came).

I just explained why random wasn't seeded. To repeat, random is only used in random tests, not in Facts.

We could use the random teats with a fixed seed in actions, or take them out into a separate project. Either is fine by me

dragon7307 commented:
Yeah, agreed. Using them with a random seed won't hurt, unless they require an abnormal amount of time to run (> 1-2 min ). 
To answer your previous question: the best way to pass a seed is probably by cli-argument, at  least for integration with actions.

Guiorgy commented:
I've renamed the `RandomData` classes to `MonkeyTests` (["Monkey testing is usually implemented as random, automated unit tests. ... Monkey testing is an effective way to identify some out-of-the-box errors"](https://en.wikipedia.org/wiki/Monkey_testing).

To avoid running them in GitHub actions, you should exclude tests that contain `MonkeyTests` from running, which can be done using the `--filter` argument:

`dotnet test --no-build --verbosity normal --filter FullyQualifiedName!~MonkeyTests`

Note: According to the documentation, on some Linux shells the `!` needs to be escaped:

> you have to escape it in some Linux or macOS shells by putting a backslash in front of it (\!)

dragon7307 commented:
Yeah, sounds good. Though it may be worth considering putting these two different types of tests in two different projects, as I can see few scenarios in which one would need to call both of them. Most of the time you would want to either prevent regression or discover new bugs - not both. I do believe that would make the project a whole lot cleaner. One would of course have the tests depend on a shared library. However, we can leave it as is for now, and see about that later.
 
Also as a side note (the same wikipedia article mentions): 
> While monkey testing is sometimes treated the same as [fuzz testing](https://en.wikipedia.org/wiki/Fuzz_testing)[[5]](https://en.wikipedia.org/wiki/Monkey_testing#cite_note-5) and the two terms are usually used together,[[6]](https://en.wikipedia.org/wiki/Monkey_testing#cite_note-6) some believe they are different by arguing that monkey testing is more about random actions while fuzz testing is more about random data input.

_Fuzz testing_ is the terminology I know and I guess it would rather apply to our tests, since '_random data input_' is what we are really doing , but I do like the idea of the new name, so I see nothing wrong with it.

But great work though!!

Guiorgy commented:
Nice catch. I know monkey tests from the time I was working on Android apps. It was a tests that would input random actions trying to crash the app. I'll change to `FuzzTests` if that's ok with you.

As for the separate projects, I started doing that, but realized that, since the integration tests are static (no random data), there's no real reason to use most of the helper methods that I wrote, which were mostly for fuzz tests and generation of output data for later reproduction. I think the only common code between the two is the `ToSystemEnumerableExtensions` which is small enough that can be duplicated in the two projects for now. Therefore, right now I am simplifying the `Facts` tests, for example:

```cs
[Fact]
public void DelimiterAtTheEndResultInEmptySpan()
{
    Assert.Equal(
        [['a', 'a'], []],
        "aab".AsSpan().Split('b').ToSystemEnumerable()
    );
}
```

Instead of the old:

```cs
[Fact]
public void DelimiterAtTheEndResultInEmptySpan()
{
    const string charArray = "aab";
    ReadOnlySpan<char> charSpan = charArray;
    const char charDelimiter = 'b';
    AssertMethodResults(
        expected: Split(charArray, charDelimiter),
        actual: charSpan.Split(charDelimiter).ToSystemEnumerable(),
        source: charArray,
        method: nameof(ReadOnlySpanExtensions.Split),
        args: ("delimiter", charDelimiter)
    );
}
```

An interesting thing I found about XUnit is that it seems to truncate some enumerable outputs, for example:

```cs
Assert.Equal(
    [['a', 'a'], ['a', 'a', 'b', 'a', 'a']],
    "aabaacaa".AsSpan().SplitAny(['b', 'c'], 2).ToSystemEnumerable()
);
```

Produces: `Expected: <>z__ReadOnlyArray<IEnumerable<char>> [[···], [···]]`

Wheres

```cs
Assert.Equal(
    (List<char[]>)[['a', 'a'], ['a', 'a', 'b', 'a', 'a']],
    "aabaacaa".AsSpan().SplitAny(['b', 'c'], 2).ToSystemEnumerable()
);
```

Produces: `Expected: [['a', 'a'], ['a', 'a', 'b', 'a', 'a']]` ¯\\\_(ツ)\_/¯

After that I will wplit the projects. I'm planning to put it into a `fuzz-tests` directory, and call the project `FuzzTests`. Do tell me if this is not quite what you want.

Guiorgy commented:
For the tests data for fuzz tests, how would you prefer me do it?

Currently I create separate `FuzzingData.cs` static class with classes that implement `IEnumerable<object[]>`:

```cs
[Theory]
[ClassData(typeof(SplitData))]
```

Would you prefer the data be inlined?

```cs
[Theory]
[InlineData(...)]
[InlineData(...)]
[InlineData(...)]
[InlineData(...)]
```

Or have members directly above the tests cases?

```cs
public static IEnumerable<object[]> SplitData() {...}

[Theory]
[MemberData(nameof(SplitData))]
```

dragon7307 commented:
That depends heavily on how you plan to structure your files. 
If you have all of the split methods in one file, then do all of that with `[ClassData]`. 
I have started to divide big files into small files, which contain one method each, as ReadOnlySpanExtensions.Linq.cs got out of hand, reaching over 2000 lines of code.  So if we want to follow this pattern with the tests, I would recommend using `[MemberData]`. `[InlineData]` expects constant values, so that shouldn't work either way with random input values.

Furthermore I would restructure the project to avoid having many root folders so putting the unit tests and the fuzz tests, maybe calling them deterministic and fuzz tests, into one folder called tests. I would also adjust the namespaces to be somewhat like 'SpanExtensions.Tests.[type of the test]'.  

Apart from that, the tests look good to me now!!! 


If you say 'take a while', how long are we talking about?

Guiorgy commented:
If we want to split the files, how about making the root `ReadOnlySpanSplitTests.cs` and such a partial class, and move the tests group classes inside it into separate files? If we do that, I'll just use local `[memberData]` as you mentioned.

How about calling the namespaces `SpanExtensions.Tests.IntegrationTests` and `SpanExtensions.Tests.FuzzTests` or `SpanExtensions.Tests.FuzzingTests`, or maybe `SpanExtensions.Tests.Integration` and `SpanExtensions.Tests.Fuzz` or just `SpanExtensions.Tests.Fuzzing` to avoid "Tests" twice?

As for the folder structure, how about this:
- solution root
  - tests
    - integration (with the Integration Tests project)
    - fuzzing (with the Fuzzing Tests project)

As discussed previously, I've split the projects into two. We could also recombine them, though most of the things needed for fuzzing (99%) are unnecessary in integration tests, so...

Let me run them and see how long they take...

dragon7307 commented:
Yeah, I intend to do something like this with all the classes - take a look at 'dev' maybe. 
Folder structure looks good.    
About the naming:   
The Unit tests (non-random ones ) are not really integration tests, are they. They are unit tests, all testing a rather small aspect of the code, that does not rely on interdependence..

Guiorgy commented:
Then how about `SpanExtensions.Tests.UnitTests` and `SpanExtensions.Tests.Fuzzing`?

dragon7307 commented:
Yeah, ok

Guiorgy commented:
While I'm doing that, from what I read xUnit only parallelizes tests methods in different classes (collections). I'm debating if I should put each `[Theory]` fuzzing method into a separate class, for example:

- class ReadOnlySpanSplits
  - class Linq (todo)
  - class Split
    - class Split
      - [Theory] Fuzz
    - class SplitWithCount
      - [Theory] Fuzz
    - class SplitWithDelimiterSequence
      - [Theory] Fuzz
    - class SplitWithDelimiterSequenceAndCount
      - [Theory] Fuzz

Does it make sence to do so?

dragon7307 commented:
Well, I do not see, how these two things relate, could you elaborate on that??  
xUnit parallelizes different classes by default, even though this is configurable behaviour. But the classes for the data don't contain any tests. So where they are stored, does not affect parallelization, does it not ???

Guiorgy commented:
I am not talking about data. The current structure is as follows:

```cs
class ReadOnlySpanSplits
{
  class Split
  {
    [Theory]
    void FuzzSplitWithCount(...){...}

    [Theory]
    void FuzzSplitWithDelimiterSequence(...){...}
    ...
  }
}
```

I wonder if it would be beneficial to change it to the following:

```cs
class ReadOnlySpanSplits
{
  class Split
  {
    class SplitWithCount
    {
        [Theory]
        void Fuzz(...){...}
    }
    class SplitWithDelimiterSequence
    {
        [Theory]
        void Fuzz(...){...}
    }
    ...
  }
}
```

This way `FuzzSplit`, `FuzzSwplitWithCount`, `FuzzSplitWithDelimiterSequence` and `FuzzSplitWithDelimiterSequenceAndCount` can all run in parallel?

dragon7307 commented:
Ah, I see, thanks for the clarification.     
I guess at the end of the day, this doesn't really make a difference, since we are going to have many tests. The degree of parallelism as enabled by the processor cores is limited by their number; having more test classes that take roughly equally long to run defeats the purpose of parallelization, since the number of the test classes will likely be higher than the number of cores available.   
So I would make the decision based upon the maintainability of the tests. Many small classes may be easier to maintain and conform more to the spirit of unit testing. I do not believe that there will be a massive difference in performance, especially as the tests grow, but the only way to truly know is to measure the performance of both.
As already stated that is probably inferior to maintainability in the case of tests that are not even run into a CI pipeline...

Guiorgy commented:
I moved the projects into the same root directory, and split tests into separate partial files (see commit [split fuzzing tests into separate files](https://github.com/draconware-dev/SpanExtensions.Net/pull/9/commits/fbd5773b8b7b6c5ec3b41b854d090d3d59bd9bab)). Tell me if that look ok to you.

I also changed to `MemberData` parameters (see commit [parameterized fuzzing tests (tmp)](https://github.com/draconware-dev/SpanExtensions.Net/pull/9/commits/23917579dda3b997a4766477f2be6bc4b724aa91)). Also tell me if that looks ok with you.

And finally, I tried separating each fuzzing test into a separate class for parallelism (see commit [separate fuzzing tests into separate classes for parallel execution (tmp)](https://github.com/draconware-dev/SpanExtensions.Net/pull/9/commits/67dfff5688c44413d48f0a4c22b05bd1bfd45ee1)). Tell me if it looks bad for maintenance. (PS. I have a 16 core, so the more parallel the better for me at least).

Also, I'm considering further separating `Split(Span<T>, T)` and `Split(Span<T>, ReadOnlySpan<T>)` into separate files like the rest, since the two have different parameter data. Right now, that file has 2 member data generators. Let me know if that's ok with you.

Guiorgy commented:
I think I'm done with the fuzzing tests for `Split` extensions.

dragon7307 commented:
Yeah, merging them as soon as they are done is my plan. 
It may really be good to focus on the source generators first.

Guiorgy commented:
I am almost done, I think. However, I just realized that I missed unit tests for Sequence Delimiter variants. I'll add them tomorow.

Guiorgy commented:
I'm done with Fuzzing and Unit Tests for Split extensions.

A few things to note:
- Currently (assuming #11 is merged), there are 2 tests that fail:
  - When an empty char array is passed to `string.Split` any whitespace character is used as the delimiter. We currently don't yet do that. The tests could temporarity be dissabled either by commenting it out, or by adding a filter to the `dotnet test` line in the actions configuration.
  - I also added tests to ensure the extensions returns the right type, i.e. `ReadOnlySpan.Split` returns `ReadOnlySpan` and `Span.Split` returns `Span`. This is currently not the case. Until the source generators are done we could either keep them as is, or, as I mentioned above, temporarily disable them.
- The names of `CountExceedingBehaviour` in this PR, including not only code, but also test names and comments, are still the old ones (`AppendLastElements`). Before we merge this, we should decide their final naming, I'll then fix them in both PRs to make it easier to merge.

dragon7307 commented:
merged into review-tests!!

dragon7307 commented:
Well the merge conflicts were quite huge and I didn't want to break anything while refactoring so i created an extra branch off of dev as an additional security layer. I have already swapped the parameters and have also adapted the code style on my local copy. I am just running a visual studio code cleanup as is configured on my device. 
I have indeed run into the issue with the infinite loops already and had my PC freeze out.

dragon7307 commented:
Furthermore, I am also currently questioning whether we actually need the tests that verify the type of Span returned, as this should not be a unit test, but is part of static analysis. 
Also that is a one time fix. So I don't believe this is a scenario for unit tests - after all, the statically enabled warnings exist for a reason

Guiorgy commented:
> already and had my PC freeze out.

oops, sorry 😅

>  I am also currently questioning whether we actually need the tests that verify the type of Span returned

If you believe they are not needed, you are free to remove them. However, keep in mind that the cast of `Span` to `ReadOnlySpan` is implicit, as such, a problem like that could escape a developer, and not everyone even looks at warnings. I included them just in case a future PR contributor made a mistake in copy-pasting code and forgot to change some types, though it's up to you.

dragon7307 commented:
I have commented them out for now!!

dragon7307 commented:
Also is there a reason options is not used in the following foreach loop?? 
```cs 
[Fact]
public void WhiteSpaceCharactersAssumedWhenDelimitersSpanIsEmpty()
{
    foreach(StringSplitOptions options in stringSplitOptions)
    {
        AssertEqual(
            [['a'], ['b'], ['c'], ['d']],
            "a b c d".AsSpan().SplitAny([], StringSplitOptions.None).ToSystemEnumerable(maxCount: 100)
        );
    }
}
```

Guiorgy commented:
Nope, that's an oversight 😅
